controller/api/public/grpc_server.go

```diff
@@ -72,9 +72,23 @@ func (s *grpcServer) ListPods(ctx context.Context, req *pb.ListPodsRequest) (*pb
 	// report from that instance and its process start time
 	reports := make(map[string]podReport)
 
+	if req.GetNamespace() != "" && req.GetSelector() != nil {
+		return nil, fmt.Errorf("cannot set both namespace and resource in the request. These are mutually exclusive")
+	}
+
 	nsQuery := ""
+	namespace := ""
 	if req.GetNamespace() != "" {
-		nsQuery = fmt.Sprintf("namespace=\"%s\"", req.GetNamespace())
+		namespace = req.GetNamespace()
+	}
+	if req.GetSelector() != nil && req.GetSelector().GetResource().GetNamespace() != "" {
```

Comment:
The intent of the new "selector" field is to filter the list of pods by their owner. For example, if I issue the following `ListPods` request, I'd expect to only get back pods owned by the emoji deployment:

```
req := &pb.ListPodsRequest{
		Selector: &pb.ResourceSelection{
			Resource: &pb.Resource{
				Namespace: "emojivoto",
				Type:      "deployment",
				Name:      "emoji",
			},
		},
	}
```

Maybe we could add a comment to the protobuf definition that indicates the intent of the "selector" field? Specifically, if "selector.resource" is present, it filters by owner resource, whereas if "selector.label_selector" is present, it filters by pod labels (see my other comment about label selectors).

We can't query the kubernetes API by owner, but we're already populating the "owner" field for all pods that are returned by the API, and we can do the filtering there. Something like:

```diff
diff --git a/controller/api/public/grpc_server.go b/controller/api/public/grpc_server.go
index 6026ce1a..1740946d 100644
--- a/controller/api/public/grpc_server.go
+++ b/controller/api/public/grpc_server.go
@@ -76,16 +76,16 @@ func (s *grpcServer) ListPods(ctx context.Context, req *pb.ListPodsRequest) (*pb
 		return nil, fmt.Errorf("cannot set both namespace and resource in the request. These are mutually exclusive")
 	}
 
+	targetOwner := req.GetSelector().GetResource()
+
 	nsQuery := ""
 	namespace := ""
 	if req.GetNamespace() != "" {
 		namespace = req.GetNamespace()
-	}
-	if req.GetSelector() != nil && req.GetSelector().GetResource().GetNamespace() != "" {
-		namespace = req.GetSelector().GetResource().GetNamespace()
-	}
-	if req.GetSelector() != nil && req.GetSelector().GetResource().GetType() == pkgK8s.Namespace {
-		namespace = req.GetSelector().GetResource().GetName()
+	} else if targetOwner.GetNamespace() != "" {
+		namespace = targetOwner.GetNamespace()
+	} else if targetOwner.GetType() == pkgK8s.Namespace {
+		namespace = targetOwner.GetName()
 	}
 	if namespace != "" {
 		nsQuery = fmt.Sprintf("namespace=\"%s\"", namespace)
@@ -124,6 +124,17 @@ func (s *grpcServer) ListPods(ctx context.Context, req *pb.ListPodsRequest) (*pb
 			continue
 		}
 
+		ownerKind, ownerName := s.k8sAPI.GetOwnerKindAndName(pod)
+		if targetOwner.GetNamespace() != "" && targetOwner.GetNamespace() != pod.GetNamespace() {
+			continue
+		}
+		if targetOwner.GetType() != "" && targetOwner.GetType() != ownerKind {
+			continue
+		}
+		if targetOwner.GetName() != "" && targetOwner.GetName() != ownerName {
+			continue
+		}
+
 		updated, added := reports[pod.Name]
 
 		status := string(pod.Status.Phase)
@@ -160,7 +171,6 @@ func (s *grpcServer) ListPods(ctx context.Context, req *pb.ListPodsRequest) (*pb
 			ProxyVersion:        proxyVersion,
 		}
 
-		ownerKind, ownerName := s.k8sAPI.GetOwnerKindAndName(pod)
 		namespacedOwnerName := pod.Namespace + "/" + ownerName
 
 		switch ownerKind {
```

Reply:
I thought about adding owner filtering in the next step once we merge this. I tend to like to keep my PRs small. But I can make that change in this PR if it makes more sense to you...